File: P2P-cpp/Graph.cpp
#ifndef GRAPH_H
#define GRAPH_H

#include "Node.cpp"
#include <unordered_map>
#include <unordered_set>
#include <string>
#include <random>

class Graph {
public:
    std::unordered_map<int, Node*> nodes;
    std::unordered_map<int, Node*> outside_nodes;
    std::unordered_map<int, Node*> all_nodes;
    int node_count;

    Graph(const std::unordered_map<int, std::vector<int>>& ADJ, const std::unordered_map<int, Node*>& node_map) {
        for (const auto& [node_id, node] : node_map) {
            nodes[node_id] = node;
            all_nodes[node_id] = node;
        }

        for (const auto& [node_id, neighbors] : ADJ) {
            for (int adjacent_id : neighbors) {
                nodes[node_id]->add_edge(node_map.at(adjacent_id));
                if (node_map.find(adjacent_id) == node_map.end()) {
                    outside_nodes[adjacent_id] = nodes[node_id]->adj[adjacent_id];
                    all_nodes[adjacent_id] = node_map.at(adjacent_id);
                }
            }
        }

        node_count = nodes.size();
    }

    std::string to_string() const {
        std::string result;
        for (const auto& [id, node] : nodes) {
            result += node->to_string() + " -> { ";
            for (const auto& adj_node : node->adj) {
                result += adj_node.second->to_string() + " ";
            }
            result += "}\n";
        }
        return result;
    }

    std::pair<std::unordered_map<int, int>, std::unordered_map<int, int>> random_walk(int source_id, int count, double alpha = 0.2) {
        Node* source_node = nodes[source_id];
        std::unordered_map<int, int> end_walk;
        std::unordered_map<int, int> escaped_walk;

        for (int i = 0; i < count; ++i) {
            Node* current_node = source_node;
            while (true) {
                if (current_node->manager != source_node->manager) {
                    escaped_walk[current_node->id]++;
                    break;
                }
                if (static_cast<double>(rand()) / RAND_MAX < alpha) {
                    end_walk[current_node->id]++;
                    break;
                }
                current_node = current_node->get_random_adjacent();
            }
        }

        return { end_walk, escaped_walk };
    }
};

#endif // GRAPH_H

File: P2P-default/Graph.py
from Node import *
import random
from queue import Queue

class Graph():
    def __init__(self, ADJ, nodes):
        self.nodes = dict()
        self.outside_nodes = dict()
        self.all_nodes = dict()

        for node_id in ADJ.keys():
            self.nodes[node_id] = nodes[node_id]
            self.all_nodes[node_id] = nodes[node_id]
            for adjacent_id in ADJ[node_id]:
                self.nodes[node_id].add_edge(nodes[adjacent_id])
                if adjacent_id not in ADJ.keys():
                    self.outside_nodes[adjacent_id] = self.nodes[node_id].adj[adjacent_id]
                    self.all_nodes[adjacent_id] = nodes[adjacent_id]

        self.node_count = len(self.nodes)

    def __repr__(self):
        adj = dict()
        for node in self.nodes.values():
            adj[str(node)] = set()
            for adj_node in node.adj:
                adj[str(node)].add(str(adj_node))
        return str(adj)

    def random_walk(self, source_id, count, alpha=0.2):
        source_node = self.nodes[source_id]
        executer = source_node.manager
        end_walk = dict()
        escaped_walk = dict()

        for i in range(count):
            current_node = source_node
            while True:
                if current_node.manager != executer:
                    escaped_walk[current_node.id] = escaped_walk.get(current_node.id, 0) + 1
                    break
                if random.random() < alpha:
                    end_walk[current_node.id] = end_walk.get(current_node.id, 0) + 1
                    break
                current_node = current_node.get_random_adjacent()

        return end_walk, escaped_walk
